server/src/main/java/org/apache/druid/metadata/segment/cache/HeapMemorySegmentMetadataCache.java

```diff
@@ -552,7 +554,13 @@ private long syncWithMetadataStore()
     updateUsedSegmentPayloadsInCache(datasourceToSummary);
     retrieveAllPendingSegments(datasourceToSummary);
     updatePendingSegmentsInCache(datasourceToSummary, syncStartTime);
-    retrieveAllSegmentSchemas(datasourceToSummary);
+
+    if (syncFinishTime.get() == null) {
+      retrieveUsedSegmentSchemasUpdatedAfter(DateTimes.COMPARE_DATE_AS_STRING_MIN, datasourceToSummary);
+    } else {
+      retrieveUsedSegmentSchemasUpdatedAfter(syncStartTime, datasourceToSummary);
```

Comment:
There should be some slack in this, to allow for the fact that clocks may not be perfectly synced across servers, and various factors (such as retries on insert) can cause records with timestamps in the past to appear. An hour should be more than enough.

Reply:
Makes sense.

There is another bug here, anyway I should have been using the start time of the previous sync rather than the current sync.

Reply:
> (such as retries on insert)

Minor clarification on this point:
For the most part, in `IndexerSQLMetadataStorageCoordinator`, I have tried to ensure that each transaction retry uses a fresh timestamp, since retries can go on for a while.
But as you point out, there can still be cases where past records appear.

P.S: I think we should be able to do something similar for fetching used segment IDs too.
Currently, we fetch all of them. But we could potentially fetch only the recently updated ones,
thus improving the delta sync time even further.
We would just need an index on `used` + `used_status_last_updated` in `druid_segments` table
(same as schemas table already does).